Failure removing an offer due to inconsistent connection details

### Description

When running the Juju Terraform Provider test `TestAcc_ResourceOffer`, we are consistently seeing, both in CI and locally, the following error during teardown,
```
        Unable to delete offer, got error: deleting offer
        "e327ccda-37db-454f-8988-d7291e195f49": cannot delete offer
        {"e327ccda-37db-454f-8988-d7291e195f49"}, it has 1 connections
--- FAIL: TestAcc_ResourceOffer (118.35s)
```

This error is unexpected because the provider logic for deleting an offer is below,
```go
	client := applicationoffers.NewClient(conn)
	offer, err := client.ApplicationOffer(ctx, input.OfferURL)
	if err != nil {
		return err
	}

	forceDestroy := false
	//This code loops until it detects 0 connections in the offer or 3 minutes elapses
	if len(offer.Connections) > 0 {
		end := time.Now().Add(5 * time.Minute)
		c.Tracef(fmt.Sprintf("offer %q has %d connections, waiting for them to be removed before destroying", offer.OfferURL, len(offer.Connections)))
		for ok := true; ok; ok = len(offer.Connections) > 0 {
			//if we have been failing to destroy offer for 5 minutes then force destroy
			//TODO: investigate cleaner solution (acceptance tests fail even if timeout set to 20m)
			if time.Now().After(end) {
				forceDestroy = true
				break
			}
			time.Sleep(10 * time.Second)
			offer, err = client.ApplicationOffer(ctx, input.OfferURL)
			if err != nil {
				return err
			}
		}
	} else {
		c.Tracef(fmt.Sprintf("offer %q has no connections, destroying", offer.OfferURL)) // <- Added for debugging locally
	}

	err = client.DestroyOffers(ctx, forceDestroy, input.OfferURL)
	if err != nil {
		return err
	}
```

Above we can see that the provider calls `client.ApplicationOffer()` until there are 0 connections against the offer.

The test has 2 main parts/steps, the first creates a model, application and offer.
The second part creates a two models, two apps (one in each model), an offer and a CMR.
At the end of the test everything is removed.

Other tests that create a CMR work fine but this test seems to fail because of the additional first step. If I remove the first step (i.e. the initial model, app and offer), the test succeeds. To be clear, the model in the first step has nothing to do with the models in the second. 

The problem roughly seems to be that during the destroy, we wait for the first model to be removed which takes some time as we wait for the pending machine to become ready. During that time the two models involved in the CMR are able to progress through their lifecycle and when they are eventually torn-down, we hit the issue with the offer connection count.

I've tested a variety of scenarios to try and debug the issue and found that if I run `juju list-offers --format yaml` repeatedly in the background (I run it against the offering model) while the test runs, the test seems to pass, granted I have a small sample size. I initially ran this script to try and capture the connection count to understand the issue better but strangely started seeing the test pass.
```bash
#!/bin/bash

set -u

output_file="${1:-./juju-list-offers.log}"

while true; do
    {
        printf '=== %s ===\n' "$(date -Iseconds)"
        if ! juju list-offers --format yaml; then
            printf 'juju list-offers failed\n'
        fi
        printf '\n'
    } >> "$output_file" 2>&1

    sleep 2
done
```

### Juju version

4.0.6-33528f3 (4.0/edge snap channel)

### Cloud

LXD

### Expected behaviour

If a call to ApplicationOffer() returns 0 connections, then a call to DestroyOffer() should succeed.

### Reproduce / Test

I was unable to replicate the issue with the Juju CLI but here was my rough attempt,
```bash
#!/bin/bash

juju add-model base
juju deploy juju-qa-dummy-source
juju offer dummy-source:sink

juju add-model source
juju deploy juju-qa-dummy-source
juju offer dummy-source:sink
juju add-model dest
juju deploy juju-qa-dummy-sink
juju relate dummy-sink admin/source.dummy-source

# Destroy the base model
# This will block until the machine has finished
# provisioning and since takes a while, allows the
# source and dest models to progress.
juju destroy-model base --no-prompt

# Check that the source model shows 1 connection
echo "Expected 1 connection in source model:"
juju switch source
juju status
juju list-offers --format yaml

# Continue in the dest model to remove the relation
juju switch dest
juju remove-relation dummy-sink dummy-source
# Wait for the relation to be removed
while juju status --relations | grep -q "dummy-source:sink"; do
    echo "Waiting for relation to be removed..."
    sleep 2
done
# Destroy the dest model and don't wait for it to finish.
timeout 2s juju destroy-model dest --no-prompt 

juju switch source
juju status 
# Wait for the offer to show 0 connections
while juju list-offers --format yaml | yq -e '[.[] | (.connections // [] | length)] | map(select(. > 0)) | length > 0' >/dev/null; do
    echo "Waiting for offer connections to be removed..."
    sleep 2
done

juju list-offers --format yaml

juju remove-offer dummy-source # This should fail saying offer has 1 connection
```

### Notes & References

Here is the Terraform trace level logging when running the test locally
[failed-offer-test.txt](https://github.com/user-attachments/files/26718175/failed-offer-test.txt)

And here is the Terraform trace level logging when running the test with the `juju list-offers` command running in the background, as well the logs from the `juju list-offers` script.

[successful-offer-test.txt](https://github.com/user-attachments/files/26718210/successful-offer-test.txt)
[successful-offer-test-watch-offer-logs.txt](https://github.com/user-attachments/files/26718213/successful-offer-test-watch-offer-logs.txt)

--- a/apiserver/facades/client/applicationoffers/applicationoffers.go
+++ b/apiserver/facades/client/applicationoffers/applicationoffers.go
@@ -437,13 +437,6 @@ func (api *OffersAPI) applicationOffersFromModel(
 		})
 	}
 
-	// If the required access level is *only* read, then return the results at
-	// this point without populating the offer connections. Offer connections
-	// are only relevant to users with admin access.
-	if requiredAccess == permission.ReadAccess {
-		return results, nil
-	}
-
 	// If the user is not a model admin or there are no results,
 	// return the results without populating the offer connections.
 	if !isModelAdmin || len(results) == 0 {

--- a/apiserver/facades/client/applicationoffers/applicationoffers_test.go
+++ b/apiserver/facades/client/applicationoffers/applicationoffers_test.go
@@ -1001,6 +1001,7 @@ func (s *offerSuite) TestFindApplicationOffers(c *tc.C) {
 		},
 	}
 	s.crossModelRelationService.EXPECT().GetOffers(gomock.Any(), domainFilters).Return(offerDetails, nil)
+	s.crossModelRelationService.EXPECT().GetOfferConnections(gomock.Any(), gomock.Any()).Return(nil, nil)
 
 	filters := params.OfferFilters{
 		Filters: []params.OfferFilter{
@@ -1108,6 +1109,7 @@ func (s *offerSuite) TestFindApplicationOffersAllOffers(c *tc.C) {
 		},
 	}
 	s.crossModelRelationService.EXPECT().GetOffers(gomock.Any(), []crossmodelrelationservice.OfferFilter{{}}).Return(offerDetails, nil)
+	s.crossModelRelationService.EXPECT().GetOfferConnections(gomock.Any(), gomock.Any()).Return(nil, nil)
 
 	filters := params.OfferFilters{Filters: []params.OfferFilter{{}}}
 
@@ -1521,6 +1523,7 @@ func (s *offerSuite) TestApplicationOffers(c *tc.C) {
 		},
 	}
 	s.crossModelRelationService.EXPECT().GetOffers(gomock.Any(), domainFilters).Return(offerDetails, nil)
+	s.crossModelRelationService.EXPECT().GetOfferConnections(gomock.Any(), gomock.Any()).Return(nil, nil)
 	args := params.OfferURLs{
 		OfferURLs: []string{"fred@external/test-model.hosted-db2", "fred@external/test-model.testing"},
 	}
@@ -1614,6 +1617,7 @@ func (s *offerSuite) TestApplicationOffersMixSuccessAndFail(c *tc.C) {
 		},
 	}
 	s.crossModelRelationService.EXPECT().GetOffers(gomock.Any(), domainFilters).Return(offerDetails, nil)
+	s.crossModelRelationService.EXPECT().GetOfferConnections(gomock.Any(), gomock.Any()).Return(nil, nil)
 	args := params.OfferURLs{
 		OfferURLs: []string{"fred@external/test-model.hosted-db2:endpoint", "fred@external/test-model.testing"},
 	}
@@ -1683,6 +1687,87 @@ func (s *offerSuite) TestApplicationOffersNotFound(c *tc.C) {
 	c.Check(obtainedOffers.Results[0].Error, tc.DeepEquals, &params.Error{
 		Message: `application offer "fred@external/test-model.testing"`,
 		Code:    params.CodeNotFound,
+	})
+}
+
+// TestApplicationOffersWithConnections tests that connections are populated
+// for admin users when querying a single offer by URL.
+func (s *offerSuite) TestApplicationOffersWithConnections(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	offerAPI := s.offerAPI(c)
+	adminTag := s.setupAuthUser("admin")
+	s.expectEntityHasPermission(adminTag, permission.SuperuserAccess)
+	adminUser := user.User{DisplayName: "fred smith"}
+	s.accessService.EXPECT().GetUserByName(gomock.Any(), user.NameFromTag(adminTag)).Return(adminUser, nil)
+
+	modelName := "prod"
+	modelOwnerTag := names.NewUserTag("fred@external")
+
+	foundModel := model.Model{
+		Name:      modelName,
+		Qualifier: model.Qualifier(modelOwnerTag.Id()),
+		UUID:      tc.Must0(c, model.NewUUID),
+	}
+	s.modelService.EXPECT().GetModelByNameAndQualifier(gomock.Any(), modelName, foundModel.Qualifier).Return(foundModel, nil)
+
+	offerUUID := uuid.MustNewUUID().String()
+	consumerModelUUID := uuid.MustNewUUID().String()
+
+	charmLocator := charm.CharmLocator{
+		Name:         "app",
+		Revision:     42,
+		Source:       charm.CharmHubSource,
+		Architecture: architecture.AMD64,
+	}
+	offerDetails := []*crossmodelrelation.OfferDetail{
+		{
+			OfferUUID:              offerUUID,
+			OfferName:              "hosted-db2",
+			ApplicationName:        "test-app",
+			ApplicationDescription: "testing application",
+			CharmLocator:           charmLocator,
+			Endpoints: []crossmodelrelation.OfferEndpoint{
+				{Name: "db"},
+			},
+			OfferUsers: []crossmodelrelation.OfferUser{{Name: "admin", Access: permission.AdminAccess}},
+		},
+	}
+	domainFilters := []crossmodelrelationservice.OfferFilter{{OfferName: "hosted-db2"}}
+	s.crossModelRelationService.EXPECT().GetOffers(gomock.Any(), domainFilters).Return(offerDetails, nil)
+
+	connections := []crossmodelrelation.OfferConnectionDetail{
+		{
+			OfferUUID:       offerUUID,
+			SourceModelUUID: consumerModelUUID,
+			RelationID:      42,
+			Username:        "consumer-user",
+			Endpoint:        "db",
+			Status:          "joined",
+			IngressSubnets:  []string{"10.0.0.0/24"},
+		},
+	}
+	s.crossModelRelationService.EXPECT().GetOfferConnections(gomock.Any(), []string{offerUUID}).Return(connections, nil)
+
+	args := params.OfferURLs{
+		OfferURLs: []string{"fred@external/prod.hosted-db2"},
+	}
+
+	obtained, err := offerAPI.ApplicationOffers(c.Context(), args)
+
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(obtained.Results, tc.HasLen, 1)
+	c.Assert(obtained.Results[0].Result, tc.NotNil)
+	c.Assert(obtained.Results[0].Result.Connections, tc.HasLen, 1)
+	c.Check(obtained.Results[0].Result.Connections[0], tc.DeepEquals, params.OfferConnection{
+		SourceModelTag: names.NewModelTag(consumerModelUUID).String(),
+		RelationId:     42,
+		Username:       "consumer-user",
+		Endpoint:       "db",
+		Status: params.EntityStatus{
+			Status: status.Status("joined"),
+		},
+		IngressSubnets: []string{"10.0.0.0/24"},
 	})
 }
 

